libsolidity/formal/CHC.cpp

```diff
@@ -654,6 +688,91 @@ vector<VariableDeclaration const*> CHC::stateVariablesIncludingInheritedAndPriva
 	return stateVars;
 }
 
+void CHC::createArrayLengths(Type const& _type)
+{
+	set<Type const*> types;
+	queue<Type const*> toVisit;
+	set<Type const*> visited;
+	toVisit.push(&_type);
+	while (!toVisit.empty())
+	{
+		auto const* type = toVisit.front();
+		toVisit.pop();
+		visited.insert(type);
+		vector<Type const*> toVisitInner;
+		if (auto arrayType = dynamic_cast<ArrayType const*>(type))
+		{
+			types.insert(type);
+			toVisitInner.emplace_back(arrayType->baseType());
+		}
+		else if (auto mapType = dynamic_cast<MappingType const*>(type))
+		{
+			toVisitInner.emplace_back(mapType->keyType());
+			toVisitInner.emplace_back(mapType->valueType());
+		}
+		else if (auto structType = dynamic_cast<StructType const*>(type))
+			toVisitInner += structType->memoryMemberTypes();
+
+		for (auto const* innerType: toVisitInner)
+			if (!visited.count(innerType))
+				toVisit.push(innerType);
+	}
+
+	auto intSort = make_shared<smt::Sort>(smt::Kind::Int);
+	for (auto const* arrayType: types)
+	{
+		solAssert(arrayType->category() == Type::Category::Array, "");
+
+		auto const& arrayTypeName = arrayType->toString(true);
+		if (m_arrayLengths.count(arrayTypeName))
+			continue;
+
+		auto arraySort = smt::smtSort(*arrayType);
+		auto lengthArraySort = make_shared<smt::ArraySort>(arraySort, intSort);
+		m_arrayLengths.emplace(
+			arrayTypeName,
+			smt::SymbolicArrayVariable(
+				lengthArraySort,
+				arrayTypeName + "_length",
+				m_context
+			)
+		);
+	}
+}
+
+void CHC::createContractArrayLengths(ContractDefinition const& _contract)
+{
+	for (auto const* var: stateVariablesIncludingInheritedAndPrivate(_contract))
+		createArrayLengths(*var->type());
+}
+
+void CHC::createFunctionArrayLengths(FunctionDefinition const& _function)
+{
+	for (auto const& var: _function.parameters())
+		createArrayLengths(*var->type());
+	for (auto const& var: _function.returnParameters())
+		createArrayLengths(*var->type());
+	for (auto const& var: _function.localVariables())
+		createArrayLengths(*var->type());
+
+	struct LengthCollector: public ASTConstVisitor
```

Comment:
Any better ideas to write this visitor? I think our lambda based generic visitor wouldn't work here

Reply:
Ref https://github.com/ethereum/solidity/issues/8315